sentry-ruby/lib/sentry/session_flusher.rb

```diff
@@ -32,6 +33,7 @@ def add_session(session)
       return unless @release
 
       ensure_thread
+      return if @exited
```

Comment:
Why don't we put this line before `ensure_thread` then we won't need to modify it? Or is there something I missed?

Reply:
`exited` can be set in 2 cases - kill and if thread creation fails with `ThreadError`. This check needs to be there for the latter.

Reply:
I think only `add_session` should care about the exit status, not the `ensure_thread`. If it's already in exit mode, `ensure_thread` shouldn't even be called. This means we need to move the `@exited` check before it.

And since `ThreadError` is an exception, I think it can have its own early return. So the flow will have 2 return points:

1. For the status check
2. For the exception

Therefore, I want to propose this instead:

```rb
    def add_session(session)
      return if @exited
      return unless @release

      begin
        ensure_thread
      rescue ThreadError
        log_debug("Session flusher thread creation failed")
        @exited = true
        return
      end

      return unless Session::AGGREGATE_STATUSES.include?(session.status)
      @pending_aggregates[session.aggregation_key] ||= init_aggregates(session.aggregation_key)
      @pending_aggregates[session.aggregation_key][session.status] += 1
    end

    # .....

    # remain unchanged
    def ensure_thread
      return if @thread&.alive?

      @thread = Thread.new do
        loop do
          sleep(FLUSH_INTERVAL)
          flush
        end
      end
    end
```

IMO, this makes the event priority more clear. Wdyt?

Reply:
yep that's much cleaner, thx!